### PR TITLE
Implement api for creating of object reference

### PR DESCRIPTION
**Pull request by bitfriend**
- Add space/convo invitation as reference to slide draft

### CODE REVIEW

File: native/acter/src/api/news.rs [hunk 1]
```diff
@@ -744,17 +747,100 @@ impl Space {
...
-            content,
-            slides: vec![],
+    pub fn new_task_ref(
+        &self,
+        target_id: String,
+        room_id: Option<String>,
+        task_list: String,
+        action: String,
+    ) -> Result<RefDetails> {
+        let target_id = EventId::parse(target_id)?;
```
Comment (gnunicorn) [inline native/acter/src/api/news.rs:757]: Why are these on `Space` if they are not related to the same space?

**Review by gnunicorn — CHANGES_REQUESTED** [review]
Looking at this code I wonder if this is the best API or if we might instead what to have a `RefDetailsBuilder`...

in either case, the functions shouldn't be bound to the `Space` if they don't need the `Space`-state at all.

File: native/acter/src/api/news.rs (hunk 1 above)
Comment (gnunicorn) [reply native/acter/src/api/news.rs:757]: just globally is fine.

Comment (bitfriend) [reply native/acter/src/api/news.rs]: Fixed by https://github.com/acterglobal/a3/pull/1369/commits/7bd1e46770feaf8f8525e410a7459c622caa256d

File: native/core/src/events/common/object_reference.rs [hunk 2]
```diff
@@ -164,6 +207,336 @@ impl RefDetails {
...
+                    self.ref_details = RefDetails::Task {
+                        target_id,
+                        room_id,
+                        task_list,
+                        action,
+                    };
+                } else {
+                    error!("couldn't parse target_id for Task ref");
+                }
+            }
```
Comment (gnunicorn) [inline native/core/src/events/common/object_reference.rs:267]: why not ...
```suggestion
    pub fn target_id(&mut self, target_id: String) -> Result<()> {
        match self.ref_details.clone() {
            RefDetails::Task {
                room_id,
                task_list,
                action,
                ..
            } => {
                let target_id = EventId::parse(target_id)?;
                  self.ref_details = RefDetails::Task {
                      target_id,
                      room_id,
                      task_list,
                      action,
                  };
                }
                // ...
            }
```

**Review by gnunicorn — APPROVED** [review]
just a minor API improvement question

File: native/core/src/events/common/object_reference.rs (hunk 2 above)
Comment (bitfriend) [reply native/core/src/events/common/object_reference.rs]: Fixed by https://github.com/acterglobal/a3/pull/1369/commits/875ea6e67087ad35ac158be8385e7ca323c4c39f

File: native/core/src/events/common/rendering.rs [hunk 3]
```diff
@@ -17,6 +18,29 @@ pub enum Position {
...
+            "top-right" => Ok(Position::TopRight),
+            "center-left" => Ok(Position::CenterLeft),
+            "center-middle" => Ok(Position::CenterMiddle),
+            "center-right" => Ok(Position::CenterRight),
+            "bottom-left" => Ok(Position::BottomLeft),
+            "bottom-middle" => Ok(Position::BottomMiddle),
+            "bottom-right" => Ok(Position::BottomRight),
+            _ => Err(crate::Error::FailedToParse {
+                model_type: "Position".to_owned(),
+                msg: "couldn't parse Position".to_owned(),
```
Comment (gnunicorn) [inline native/core/src/events/common/rendering.rs:38]:
```suggestion
                msg: format!("${str} not a valid Position"),
```

File: native/core/src/events/common/object_reference.rs [hunk 4]
```diff
@@ -17,6 +18,24 @@ pub enum TaskAction {
...
+    fn from_str(s: &str) -> Result<Self, crate::Error> {
+        match s {
+            "link" => Ok(TaskAction::Link),
+            "embed" => Ok(TaskAction::Embed),
+            "embed-subscribe" => Ok(TaskAction::EmbedSubscribe),
+            "embed-accept-assignment" => Ok(TaskAction::EmbedAcceptAssignment),
+            "embed-mark-done" => Ok(TaskAction::EmbedMarkDone),
+            _ => Err(crate::Error::FailedToParse {
+                model_type: "TaskAction".to_owned(),
+                msg: "couldn't parse TaskAction".to_owned(),
```
Comment (gnunicorn) [inline native/core/src/events/common/object_reference.rs:33]:
```suggestion
                msg: format!("${str} not a valid TaskAction"),
```

File: native/core/src/events/common/object_reference.rs [hunk 5]
```diff
@@ -55,6 +90,22 @@ impl CalendarEventAction {
...
+    type Err = crate::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "link" => Ok(CalendarEventAction::Link),
+            "embed" => Ok(CalendarEventAction::Embed),
+            "embed-rsvp" => Ok(CalendarEventAction::EmbedRsvp),
+            _ => Err(crate::Error::FailedToParse {
+                model_type: "CalendarEventAction".to_owned(),
+                msg: "couldn't parse CalendarEventAction".to_owned(),
```
Comment (gnunicorn) [inline native/core/src/events/common/object_reference.rs:103]:
```suggestion
                msg: format!("${str} not a valid CalendarEventAction"),
```

**Review by gnunicorn — APPROVED** [review]
Thanks, mate. Just a quick one to put more debug info into the error message.

File: native/core/src/events/common/rendering.rs (hunk 3 above)
Comment (bitfriend) [reply native/core/src/events/common/rendering.rs]: Fixed by https://github.com/acterglobal/a3/pull/1369/commits/12557fb12865deecdf114e898a47651624c003d0

File: native/core/src/events/common/object_reference.rs (hunk 4 above)
Comment (bitfriend) [reply native/core/src/events/common/object_reference.rs]: Fixed by https://github.com/acterglobal/a3/pull/1369/commits/12557fb12865deecdf114e898a47651624c003d0

File: native/core/src/events/common/object_reference.rs [hunk 6]
```diff
@@ -39,6 +58,22 @@ impl TaskListAction {
...
+    type Err = crate::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "link" => Ok(TaskListAction::Link),
+            "embed" => Ok(TaskListAction::Embed),
+            "embed-subscribe" => Ok(TaskListAction::EmbedSubscribe),
+            _ => Err(crate::Error::FailedToParse {
+                model_type: "TaskListAction".to_owned(),
+                msg: "couldn't parse TaskListAction".to_owned(),
```
[the remark this answers is not in the document]
Comment (bitfriend) [reply native/core/src/events/common/object_reference.rs]: Fixed by https://github.com/acterglobal/a3/pull/1369/commits/12557fb12865deecdf114e898a47651624c003d0

File: native/core/src/events/common/object_reference.rs (hunk 5 above)
Comment (bitfriend) [reply native/core/src/events/common/object_reference.rs]: Fixed by https://github.com/acterglobal/a3/pull/1369/commits/12557fb12865deecdf114e898a47651624c003d0

**Pull request merged**